compiler/rustc_middle/src/ty/fold.rs

```diff
@@ -423,14 +448,18 @@ impl<'a, 'tcx> TypeFolder<'tcx> for BoundVarReplacer<'a, 'tcx> {
     fn fold_region(&mut self, r: ty::Region<'tcx>) -> ty::Region<'tcx> {
         match *r {
             ty::ReLateBound(debruijn, br) if debruijn == self.current_index => {
-                let region = (self.fld_r)(br);
+                let region = self.delegate.replace_region(br);
                 if let ty::ReLateBound(debruijn1, br) = *region {
                     // If the callback returns a late-bound region,
                     // that region should always use the INNERMOST
                     // debruijn index. Then we adjust it to the
                     // correct depth.
                     assert_eq!(debruijn1, ty::INNERMOST);
-                    self.tcx.mk_region(ty::ReLateBound(debruijn, br))
+                    if debruijn == ty::INNERMOST {
+                        region
+                    } else {
+                        self.tcx.mk_region(ty::ReLateBound(debruijn, br))
+                    }
```

Comment:
Isn't there a mk_or_reuse_region`?

Reply:
not that i know of, we previously used `mk_or_reuse_ty` during type folding but afaict that has been removed in favor of a manual check as well.

Reply:
it's called `reuse_or_mk_region` instead :grin: